GCEWindowsAgent/wsfc.go

```diff
@@ -0,0 +1,276 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-windows/logger"
+)
+
+const wsfcDefaultAgentPort = "59998"
+
+type agentState int
+
+// Enum for agentState
+const (
+	running agentState = iota
+	stopped
+)
+
+var (
+	once          sync.Once
+	agentInstance *wsfcAgent
+)
+
+type wsfcManager struct {
+	agentNewState agentState
+	agentNewPort  string
+	agent         healthAgent
+}
+
+// Create new wsfcManager based on metadata
+// agent request state will be set to running if one of the following is true:
+// - EnableWSFC is set
+// - WSFCAddresses is set (As an advanced setting, it will always override EnableWSFC flag)
+func newWsfcManager(newMetadata *metadataJSON) *wsfcManager {
+	newState := stopped
+	if newMetadata.Instance.Attributes.EnableWSFC ||
+		len(newMetadata.Instance.Attributes.WSFCAddresses) > 0 {
+		newState = running
+	}
+
+	newPort := wsfcDefaultAgentPort
+	if len(newMetadata.Instance.Attributes.WSFCAgentPort) > 0 {
+		newPort = newMetadata.Instance.Attributes.WSFCAgentPort
+	}
+
+	return &wsfcManager{agentNewState: newState, agentNewPort: newPort, agent: getWsfcAgentInstance()}
+}
+
+// Implement manager.diff()
+func (m *wsfcManager) diff() bool {
+	return m.agentNewState != m.agent.getState() || m.agentNewPort != m.agent.getPort()
+}
+
+// Implement manager.disabled().
+// wsfc manager is always enabled. The manager is just a broker which manages the state of wsfcAgent. User
+// can disable the wsfc feature by setting the metadata. If the manager is disabled, the agent will become
+// an orphan goroutine if it is currently running and no one can talk to it.
+func (m *wsfcManager) disabled() bool {
+	return false
+}
+
+// Diff will always be called before set. So in set, only two cases are possible:
+// - state changed: start or stop the wsfc agent accordingly
+// - port changed: restart the agent if it is running
+func (m *wsfcManager) set() error {
+	m.agent.setPort(m.agentNewPort)
+
+	// if state changes
+	if m.agentNewState != m.agent.getState() {
+		if m.agentNewState == running {
+			return m.startAgent()
+		}
+
+		return m.stopAgent()
+	}
+
+	// If port changed
+	if m.agent.getState() == running {
+		if err := m.stopAgent(); err != nil {
+			return err
+		}
+
+		return m.startAgent()
+	}
+
+	return nil
+}
+
+// Start health agent in goroutine
+func (m *wsfcManager) startAgent() error {
```

Comment:
originally, there is some additional logic to create channel and pass it to agent run/stop. But we no longer have it now. removed